docs/design/sum_types.md

```diff
@@ -0,0 +1,197 @@
+# Sum types
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [`choice` declarations](#choice-declarations)
+-   [User-defined sum types](#user-defined-sum-types)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+In Carbon, a _sum type_ is a type whose values are grouped into several distinct
+named cases, called _alternatives_. A value of a sum type logically consists of
+a value of one of the alternatives, together with a _discriminator_ tag, which
+identifies which alternative is present. Sum types are typically handled with
```

Comment:
I think there are extra commas here, making it hard to read: is "together with a discriminator tag" intended to be a sub-clause, with "which identifies which alternative is present" referring to the value of one of the alternatives? Or is the third part referring to the discriminator tag?

Consider rephrasing, e.g.:

...a value of one of the alternatives together with a _discriminator_ tag. The discriminator...

However, as I think about this further, is this a good explanation? Consider:

```
choice Foo {
  Bar,
  Baz,
}
```

This is only the discriminator, no additional values, right? Perhaps it would be better to say:

A value of a sum type consists of a _discriminator_ tag which identifies which alternative is present, as well as the alternative's value if it has one.